src/nunavut/lang/c/templates/_init_macros.j2

```diff
@@ -1,21 +1,29 @@
-{%- macro init_field(instance_ptr_token, field, index='') -%}
-    {%- if field is not padding -%}
-    {%- if field.data_type is CompositeType -%}
-        {{ field.data_type | declaration }}_init(&({{ instance_ptr_token }}->{{ field | id }}{{ index }}));
-    {%- elif field.data_type is FixedLengthArrayType -%}
-        {%- if field.data_type.element_type is PrimitiveType -%}
-        memset(&({{ instance_ptr_token }}->{{ field | id }}[0]), 0, sizeof({{ field.data_type | declaration }}) * {{ field.data_type.capacity }});
+{%- macro init_field(field, data_type=None, ref='out_instance->') -%}
+    {%- if field is None -%}
+        {%- set _data_type = data_type -%}
+        {%- set _out = ref -%}
+    {%- else -%}
+        {%- set _data_type = field.data_type -%}
+        {%- set _out = ref + (field | id) -%}
+    {%- endif -%}
+
+    {%- if field is not padding or field is None -%}
+    {%- if _data_type is CompositeType -%}
+        {{ _data_type | declaration }}_init(&({{ _out }}));
```

Comment:
the ref default is `out_instance->` which means if you have `init_field(None, _data_type.element_type)` this would generate `&(out_instance->)` which is invalid syntax.